test/cbmc/sources/core_json_contracts.c

```diff
@@ -0,0 +1,185 @@
+        char * endOfBuf = buf + max;
+        return pointer_in_range( buf, endOfValue, endOfBuf );
+    }
+
+    return true;
+}
+
+bool JSON_IteratePreconditions( char * buf,
+                                size_t max,
+                                size_t * start,
+                                size_t * next,
+                                JSONPair_t * outPair )
+{
+    return ( 0 < max && max < CBMC_MAX_BUFSIZE )
+           & ( buf == NULL || allocated( buf, max ) )
+           & ( start == NULL || allocated( start, sizeof( *start ) ) )
+           & ( next == NULL || allocated( next, sizeof( *next ) ) )
+           & ( outPair == NULL || allocated( outPair, sizeof( *outPair ) ) )
+           /* *INDENT-OFF* */
+           & ( outPair != NULL ==> ( ( outPair->keyLength == 0 && outPair->key == NULL ) || allocated( outPair->key, outPair->keyLength ) ) )
+           & ( outPair != NULL ==> ( ( outPair->valueLength == 0 && outPair->value == NULL ) || allocated( outPair->value, outPair->valueLength ) ) );
+           /* *INDENT-ON* */
+}
+
+JSONStatus_t JSON_ValidatePreconditions( char * buf,
+                                         size_t max )
+{
+    return ( max < CBMC_MAX_BUFSIZE )
+           & ( buf == NULL || allocated( buf, max ) );
+}
+
+bool JSON_IteratePostconditions( JSONStatus_t result,
+                                 char * buf,
+                                 size_t max,
+                                 JSONPair_t * outPair )
+{
+    CHECK( isJSONIterateEnum( result ) );
+
+    if( result == JSONSuccess )
+    {
+/* *INDENT-OFF* */
+        CHECK( ( outPair->key != NULL ) ==> ( ( outPair->key > buf ) && ( ( outPair->key + outPair->keyLength ) < ( buf + max ) ) ) );
+        CHECK( ( outPair->key != NULL ) ==> ( ( outPair->key + outPair->keyLength ) < outPair->value ) );
+/* *INDENT-ON* */
+        CHECK( ( outPair->value > buf ) && ( ( outPair->value + outPair->valueLength ) <= ( buf + max ) ) );
+        CHECK( isJSONTypesEnum( outPair->jsonType ) );
+    }
+
+    return true;
+}
+
+bool arraySearchPreconditions( char * buf,
+                               size_t max,
+                               size_t * outValue,
+                               size_t * outValueLength )
+{
+    return ( isValidBoundedBuffer( buf, max ) )
+           & ( allocated( outValue, sizeof( *outValue ) ) )
+           & ( allocated( outValueLength, sizeof( *outValueLength ) ) )
+           & ( *outValueLength <= max );
+}
+
+bool objectSearchPreconditions( char * buf,
+                                size_t max,
+                                const char * query,
+                                size_t queryLength,
+                                size_t * outValue,
+                                size_t * outValueLength )
+{
+    return arraySearchPreconditions( buf, max, outValue, outValueLength )
+           & ( queryLength < CBMC_MAX_QUERYKEYLENGTH )
+           & ( allocated( query, queryLength ) );
+}
+
```

Comment:
can we move that after `arraySearchPostconditions` so that `arraySearchPreconditions` and `arraySearchPostconditions` are next to each other?